### PR TITLE
Add Plugins for OneClick and Clavis

### PR DESCRIPTION
**Pull request by benjamingregory**
### DAG
- ClavisToExcel
- OneClickToExcel

### Plugin
- OneClickToSpreadsheet
- ClavisToSpreadsheet
- S3ToSpreadsheet

### Operator
- ClavisToS3
- OneClickToS3
- S3ToSpreadsheet

### Hook
- ClavisHook

### CODE REVIEW

File: dags/ClavisToExcel/clavis_to_excel.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    kick_off_dag = DummyOperator(task_id='kick_off_dag')
+    for endpoint in endpoints:
+        clavis_to_s3 = ClavisToS3Operator(
+            task_id=('clavis_output_{0}'.format(endpoint)),
+            http_conn_id='clavis-connection-id',
+            clavis_endpoint=endpoint,
+            payload={'report_date': '{{ yesterday_ds }}'},
+            s3_conn_id='s3-connection-id',
+            s3_bucket='bucket-name',
+            s3_key=('/path/to/file/{0}.json'.format(endpoint)))
```
Comment (andscoop) [inline dags/ClavisToExcel/clavis_to_excel.py]: I think you have extra wrapping around this - maybe it was once on multiple lines?

File: dags/ClavisToExcel/clavis_to_excel.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          schedule_interval='@once',
+          catchup=False)
+
+endpoints = ['products', 'search_terms', 'content']
+
+with dag:
+    kick_off_dag = DummyOperator(task_id='kick_off_dag')
+    for endpoint in endpoints:
+        clavis_to_s3 = ClavisToS3Operator(
+            task_id=('clavis_output_{0}'.format(endpoint)),
```
Comment (andscoop) [inline dags/ClavisToExcel/clavis_to_excel.py]: Extra wrapping with ()

File: dags/ClavisToExcel/clavis_to_excel.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            payload={'report_date': '{{ yesterday_ds }}'},
+            s3_conn_id='s3-connection-id',
+            s3_bucket='bucket-name',
+            s3_key=('/path/to/file/{0}.json'.format(endpoint)))
+
+        i = '{0}.json'.format(endpoint)
+        o = "{0}.xlsx".format(endpoint)
+
+        s3_to_excel = S3ToSpreadsheetOperator(
+            task_id=('read_{0}_into_excel'.format(endpoint)),
```
Comment (andscoop) [inline dags/ClavisToExcel/clavis_to_excel.py]: Extra wrapping

File: dags/OneClickToExcel/one_click_to_excel.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          default_args=default_args,
+          schedule_interval='@once',
+          catchup=False)
+
+with dag:
+    kick_off_dag = DummyOperator(task_id='kick_off_dag')
+
+    one_click_to_s3 = OneClickToS3Operator(
+        task_id='one_click',
+        one_click_conn='one_click-connection-id',
```
Comment (andscoop) [inline dags/OneClickToExcel/one_click_to_excel.py]: Nitpicky but I don't think we should be mixing "-" and "_"

File: plugins/ClavisToSpreadsheetPlugin/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# Plugin - Clavis To Spreadsheet
```
Comment (andscoop) [inline plugins/ClavisToSpreadsheetPlugin/README.md]: mmmm docs

File: plugins/ClavisToSpreadsheetPlugin/hooks/ClavisHook.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from airflow.hooks.http_hook import HttpHook
+
+
+class ClavisHook(HttpHook):
+    def __init__(self, method='GET', http_conn_id='http_default'):
+        super().__init__(method, http_conn_id)
+
+    def get_conn(self, headers=None):
```
Comment (andscoop) [inline plugins/ClavisToSpreadsheetPlugin/hooks/ClavisHook.py]: I like this. Clever way around your problem.

You're calling super().get_conn(headers) in both cases so I think you can clean this up to be

session = super().get_connect(headers)
session.auth = session.auth if self.endpoint == 'token' else None
return session

File: plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
+from airflow.utils.decorators import apply_defaults
+from airflow.hooks.S3_hook import S3Hook
+from airflow.models import BaseOperator
+import pandas as pd
+
+
+class S3ToSpreadsheetOperator(BaseOperator):
+    """
+    S3 to Spreadsheet Operator
```
Comment (andscoop) [inline plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py]: Doc strings look great btw

File: plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        self.output_format = output_format
+        self.output_payload = output_payload
+        self.filters = filters
+        self.append_fields = append_fields
+
+    def execute(self, context):
+        input_s3 = S3Hook(s3_conn_id=self.input_s3_conn_id)
+        output_s3 = S3Hook(s3_conn_id=self.output_s3_conn_id)
+
+        ik = (input_s3.get_key(self.input_s3_key,
```
Comment (andscoop) [inline plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py]: all your other variable names are much more explicit.

File: plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            df.to_excel(w, **self.output_payload)
+            w.save()
+            output_s3.load_file(
+                filename='temp.xlsx',
+                key=self.output_s3_key,
+                bucket_name=self.output_s3_bucket,
+                replace=True)
+            output_s3.connection.close()
+        elif self.output_format.lower() == 'csv':
+            output_string = df.to_csv(**self.output_payload)
```
Comment (andscoop) [inline plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py]: 👍

File: dags/OneClickToExcel/one_click_to_excel.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        input_file_type='csv',
+        output_s3_conn_id='output-s3-connection-id',
+        output_s3_bucket='output-bucket-name',
+        output_s3_key='/path/to/file/{{ ds }}.xlsx',
+        output_format='excel',
+        output_payload={"index": False},
+        append_fields={'run_date': '{{ ds }}'}
+    )
+
+    kick_off_dag >> one_click_to_s3 >> s3_to_excel
```
Comment (joshuathompson) [inline dags/OneClickToExcel/one_click_to_excel.py]: @andscoop did we decide to use set_downstream, set_upstream, or these bitwise operators?  Does it matter?

File: dags/ClavisToExcel/clavis_to_excel.py (hunk 3 above)
Comment (benjamingregory) [reply dags/ClavisToExcel/clavis_to_excel.py]: Fixed.

File: dags/ClavisToExcel/clavis_to_excel.py (hunk 1 above)
Comment (benjamingregory) [reply dags/ClavisToExcel/clavis_to_excel.py]: Fixed.

File: dags/ClavisToExcel/clavis_to_excel.py (hunk 2 above)
Comment (benjamingregory) [reply dags/ClavisToExcel/clavis_to_excel.py]: Fixed.

File: dags/OneClickToExcel/one_click_to_excel.py (hunk 4 above)
Comment (benjamingregory) [reply dags/OneClickToExcel/one_click_to_excel.py]: Fixed.

File: plugins/ClavisToSpreadsheetPlugin/hooks/ClavisHook.py (hunk 6 above)
Comment (benjamingregory) [reply plugins/ClavisToSpreadsheetPlugin/hooks/ClavisHook.py]: Fixed.

File: plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py (hunk 8 above)
Comment (benjamingregory) [reply plugins/ClavisToSpreadsheetPlugin/operators/S3ToSpreadsheetOperator.py]: I did that for shorter line lengths but I agree it should be more explicit. I'll fix.

File: dags/OneClickToExcel/one_click_to_excel.py (hunk 10 above)
Comment (benjamingregory) [reply dags/OneClickToExcel/one_click_to_excel.py]: Leaving as same until definitive stylistic decision made.

**Pull request merged**